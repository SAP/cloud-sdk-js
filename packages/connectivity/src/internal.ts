export * from './http-agent';
export * from './scp-cf';
