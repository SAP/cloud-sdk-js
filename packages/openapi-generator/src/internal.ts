export * from './file-serializer';
export * from './options';
export * from './parser';
export * from './sdk-metadata';
export * from './document-converter';
export * from './extensions';
export * from './generator';
export * from './openapi-types';
export * from './schema-util';
