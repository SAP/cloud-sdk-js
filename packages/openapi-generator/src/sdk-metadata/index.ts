export * from './code-sample-util';
export * from './code-sample';
export * from './generation-and-usage';
export * from './sdk-metadata';
