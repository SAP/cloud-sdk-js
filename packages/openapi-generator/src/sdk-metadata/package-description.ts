/**
 * @internal
 */
export const packageDescription = (packageName: string): string =>
  `SAP Cloud SDK for JavaScript: Generated client for OpenAPI service ${packageName}`;
