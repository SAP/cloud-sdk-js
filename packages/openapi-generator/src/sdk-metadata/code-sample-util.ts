import { getLevenshteinClosest } from '@sap-cloud-sdk/generator-common/internal';
import { codeBlock, pascalCase } from '@sap-cloud-sdk/util';
import { OpenApiApi, OpenApiOperation } from '../openapi-types';

/**
 * @internal
 */
export function getMainApi(
  serviceName: string,
  apis: OpenApiApi[]
): OpenApiApi {
  return (
    getLevenshteinClosest(serviceName, apis, x => x.name) ||
    getApiWithMaxOperations(apis)
  );
}
/**
 * @internal
 */
export function getApiWithMaxOperations(apis: OpenApiApi[]): OpenApiApi {
  const sortedByOperationsLength = apis.sort((a, b) =>
    a.operations.length > b.operations.length ? -1 : 1
  );
  const withGetAll = sortedByOperationsLength.find(api =>
    api.operations.find(operation =>
      operation.operationId.toLocaleLowerCase('getall')
    )
  );
  if (withGetAll) {
    return withGetAll;
  }
  // return the one with most methods
  return sortedByOperationsLength[0];
}
/**
 * @internal
 */
export function getMainOperation(api: OpenApiApi): OpenApiOperation {
  if (api.operations.length === 1) {
    return api.operations[0];
  }
  return (
    getLevenshteinClosest(api.name, api.operations, op => op.operationId) ||
    getGetAllOperation(api.operations) ||
    getGetOperation(api.operations) ||
    getAnyOperationWithoutParams(api.operations) ||
    api.operations[0]
  );
}
/**
 * @internal
 */
export function getGetAllOperation(
  operations: OpenApiOperation[]
): OpenApiOperation | undefined {
  return operations.find(
    operation => operation.operationId.toLowerCase() === 'getall'
  );
}
/**
 * @internal
 */
export function getGetOperation(
  operations: OpenApiOperation[]
): OpenApiOperation | undefined {
  const getOperations = operations.filter(
    operation => operation.method === 'get'
  );

  // check if get operation without parameters exists
  // else return operation with min parameters
  if (getOperations.length > 0) {
    const getOperationsWithoutParams = getOperations.filter(
      getOperation =>
        getOperation.pathParameters.length === 0 &&
        getOperation.queryParameters.length === 0
    );
    if (getOperationsWithoutParams.length > 0) {
      return getOperationsWithoutParams[0];
    }
    return getOperations[0];
  }
  return undefined;
}
/**
 * @internal
 */
export function getAnyOperationWithoutParams(
  operations: OpenApiOperation[]
): OpenApiOperation | undefined {
  return operations.find(
    operation =>
      operation.pathParameters.length === 0 &&
      operation.queryParameters.length === 0
  );
}
/**
 * @internal
 */
export function getOperationParamCode(operation: OpenApiOperation): string {
  const paramSignature: string[] = [];
  if (operation.pathParameters?.length) {
    paramSignature.push(
      operation.pathParameters.map(pathParam => pathParam.name).join(', ')
    );
  }

  if (operation.requestBody) {
    let reqBody = 'undefined';
    if (operation.requestBody.required) {
      reqBody = 'myRequestBody';
    }
    paramSignature.push(reqBody);
  }

  if (operation.queryParameters) {
    const queryParams = operation.queryParameters
      .filter(queryParam => queryParam.required)
      .map(
        queryParam => `${queryParam.name}: 'my${pascalCase(queryParam.name)}'`
      );

    if (queryParams.length > 0) {
      paramSignature.push(`{ ${queryParams.join(', ')} }`);
    }
  }

  return codeBlock`${paramSignature.join(', ')}`;
}
