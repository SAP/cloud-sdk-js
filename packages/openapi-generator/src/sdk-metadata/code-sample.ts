import { codeBlock } from '@sap-cloud-sdk/util';
import { MultiLineText } from '@sap-cloud-sdk/generator-common/internal';
import { OpenApiOperation } from '../openapi-types';
import { getOperationParamCode } from './code-sample-util';

/**
 * @internal
 */
export function apiSpecificCodeSample(
  apiName: string,
  operation: OpenApiOperation,
  directoryName: string
): MultiLineText {
  return codeBlock`
import { ${apiName} } from './generated/${directoryName}';

const responseData = await ${apiName}.${
    operation.operationId
  }(${getOperationParamCode(
    operation
  )}).execute({ destinationName: 'myDestinationName' });
`;
}
