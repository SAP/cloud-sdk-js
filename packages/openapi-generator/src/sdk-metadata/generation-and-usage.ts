import { MultiLineText } from '@sap-cloud-sdk/generator-common/internal';
import { OpenApiDocument } from '../openapi-types';
import { apiSpecificCodeSample } from './code-sample';
import { getMainApi, getMainOperation } from './code-sample-util';

/**
 * @internal
 */
export function getApiSpecificUsage(
  openApiDocument: OpenApiDocument
): MultiLineText {
  const apisWithOperations = openApiDocument.apis.filter(
    api => api.operations?.length > 0
  );

  if (apisWithOperations.length > 0) {
    const mainApi = getMainApi(openApiDocument.serviceName, apisWithOperations);
    const operation = getMainOperation(mainApi);

    const instructions = apiSpecificCodeSample(
      mainApi.name,
      operation,
      openApiDocument.serviceOptions.directoryName
    );
    return instructions;
  }
  return '';
}
