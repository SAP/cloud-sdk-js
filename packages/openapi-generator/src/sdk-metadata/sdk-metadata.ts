import {
  getSdkMetadataClient,
  Client
} from '@sap-cloud-sdk/generator-common/internal';
import { OpenApiDocument } from '../openapi-types';
import { getApiSpecificUsage } from './generation-and-usage';

/**
 * @internal
 */
export async function sdkMetadata(
  openApiDocument: OpenApiDocument
): Promise<Client> {
  const generationAndUsage = getApiSpecificUsage(openApiDocument);
  return getSdkMetadataClient(generationAndUsage, 'OpenAPI');
}
