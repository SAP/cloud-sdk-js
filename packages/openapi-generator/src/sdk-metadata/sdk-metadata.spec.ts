import { getSdkVersion } from '@sap-cloud-sdk/generator-common/internal';
import { dummyOpenApiDocument } from '../../test/test-util';
import { sdkMetadata } from './sdk-metadata';

describe('sdk-metadata', () => {
  it('generates metadata content for services without pregenerated lib', async () => {
    const metadata = await sdkMetadata(dummyOpenApiDocument);
    expect(metadata).toMatchSnapshot({
      generatorVersion: expect.any(String)
    });
    expect(metadata.generatorVersion).toEqual(await getSdkVersion());
  });
});
