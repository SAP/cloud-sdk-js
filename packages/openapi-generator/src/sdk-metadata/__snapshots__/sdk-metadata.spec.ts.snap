// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sdk-metadata generates metadata content for services without pregenerated lib 1`] = `
Object {
  "apiSpecificUsage": "import { DummyApi } from './generated/dummy-package';

const responseData = await DummyApi.DummyFunction(dummyId, undefined, { query_1: 'myQuery1' }).execute({ destinationName: 'myDestinationName' });",
  "apiType": "OpenAPI",
  "generatorVersion": Any<String>,
  "language": "JavaScript",
  "serviceStatus": "verified",
}
`;
